adapters/unicorn/unicorn.go

```diff
@@ -0,0 +1,285 @@
+		return nil, []error{err}
+	}
+
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestData := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    requestJSON,
+		Headers: getHeaders(request),
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+}
+
+func getHeaders(request *openrtb.BidRequest) http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+
+	if request.Device != nil {
+		if len(request.Device.UA) > 0 {
+			headers.Add("User-Agent", request.Device.UA)
+		}
+
+		if len(request.Device.IPv6) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IPv6)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		}
+	}
+
+	return headers
+}
+
+func setImps(request *openrtb.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]openrtb.Imp, error) {
+	for i := 0; i < len(request.Imp); i++ {
+		imp := &request.Imp[i]
+
+		var ext unicornImpExt
+		err := json.Unmarshal(imp.Ext, &ext)
+
+		if err != nil {
+			return nil, &errortypes.BadInput{
+				Message: fmt.Sprintf("Error while decoding imp[%d].ext: %s", i, err),
+			}
+		}
+
+		var placementID string
+		if ext.Bidder.PlacementID != "" {
+			placementID = ext.Bidder.PlacementID
+		} else {
+			placementID, err = getStoredRequestImpID(imp)
+			if err != nil {
+				return nil, &errortypes.BadInput{
+					Message: fmt.Sprintf("Error get StoredRequestImpID from imp[%d]: %s", i, err),
+				}
+			}
+		}
+
+		ext.Bidder.PlacementID = placementID
+
+		imp.Ext, err = json.Marshal(ext)
+		if err != nil {
+			return nil, &errortypes.BadInput{
+				Message: fmt.Sprintf("Error while encoding imp[%d].ext: %s", i, err),
+			}
+		}
+
+		secure := int8(1)
+		imp.Secure = &secure
+		imp.TagID = placementID
+	}
+	return request.Imp, nil
+}
+
+func getStoredRequestImpID(imp *openrtb.Imp) (string, error) {
+	var impExt map[string]json.RawMessage
+
+	err := json.Unmarshal(imp.Ext, &impExt)
+
+	if err != nil {
+		return "", fmt.Errorf("Error while decoding ext because: %s", err)
+	}
+
+	rawPrebidExt, ok := impExt[openrtb_ext.PrebidExtKey]
+
+	if !ok {
+		return "", fmt.Errorf("ext.prebid is null")
+	}
+
+	var prebidExt openrtb_ext.ExtImpPrebid
+
+	err = json.Unmarshal(rawPrebidExt, &prebidExt)
+
+	if err != nil {
+		return "", fmt.Errorf("cannot decoding ext.prebid because: %s", err)
+	}
+
+	if prebidExt.StoredRequest == nil {
+		return "", fmt.Errorf("ext.prebid.storedrequest is null")
+	}
+
+	return prebidExt.StoredRequest.ID, nil
+}
+
+func setSourceExt() (json.RawMessage, error) {
+	siteExt, err := json.Marshal(unicornSourceExt{
+		Stype:  "prebid_server_uncn",
+		Bidder: "unicorn",
+	})
+	if err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("Error while encoding source.ext, err: %s", err),
+		}
+	}
+	return siteExt, nil
```

Comment:
It seems wasteful to marshal this for every call when it contains all static content. Consider hardcoding a json raw message or moving the marshal to do the builder if you'd like to continue doing this dynamically.

Reply:
I use hardcoding a json raw message. fixed.